File: windows/src/Slipstream.App/Pages/BrowseViewModel.cs
using CommunityToolkit.Mvvm.ComponentModel;
using Slipstream.App.Services;
using Slipstream.Core.Files;
using Slipstream.Meridian.Controls;

namespace Slipstream.App.Pages;

/// <summary>The four filter chips spec §12 shows above the Browse data table.</summary>
public enum BrowseFilter
{
    All,
    Video,
    Audio,
    Images,
    Docs,
}

/// <summary>
/// Drives the Browse-phone page: a directories-first/alphabetical listing of one remote
/// folder, filterable by MIME-prefix chip, with breadcrumb navigation and an honest notice
/// when the peer truncated the listing (spec §6's 5000-entry cap).
/// </summary>
/// <remarks>
/// The Docs chip has no single canonical MIME prefix the way Video/Audio/Images do, so it
/// is defined as the practical complement: anything under "application/" (pdf, Word, Excel,
/// PowerPoint, zip, octet-stream, ...) or "text/" (plain text, markdown, ...). That covers
/// every non-media file a phone's shared storage is likely to hold without requiring an
/// exhaustive allowlist of document extensions.
/// </remarks>
public sealed partial class BrowseViewModel : ObservableObject
{
    private readonly IPeerHost _peerHost;

    /// <summary>Breadcrumb stack of path segments, root first. Rebuilt from the full path
    /// each navigation rather than tracked as raw remote paths, since that is what the UI
    /// renders and what NavigateBackAsync needs to rebuild a parent path.</summary>
    private readonly List<string> _pathSegments = [];

    private IReadOnlyList<FileEntry> _allEntries = [];
    private IReadOnlyList<FileEntry> _entries = [];
    private BrowseFilter _selectedFilter = BrowseFilter.All;
    private string _searchText = string.Empty;
    private IReadOnlyList<string> _breadcrumbs = [];
    private string? _truncationNotice;
    private MeridianStateViewState _state = MeridianStateViewState.Loading;
    private string? _errorMessage;

    public BrowseViewModel(IPeerHost peerHost)
    {
        ArgumentNullException.ThrowIfNull(peerHost);
        _peerHost = peerHost;
    }

    /// <summary>The current folder's entries after directories-first/alphabetical sort and
    /// the active filter chip.</summary>
    public IReadOnlyList<FileEntry> Entries
    {
        get => _entries;
        private set => SetProperty(ref _entries, value);
    }

    public BrowseFilter SelectedFilter
    {
        get => _selectedFilter;
        private set => SetProperty(ref _selectedFilter, value);
    }

    /// <summary>Path segments from root to the current folder, e.g. ["storage", "DCIM"].</summary>
    public IReadOnlyList<string> Breadcrumbs
    {
        get => _breadcrumbs;
        private set => SetProperty(ref _breadcrumbs, value);
    }

    /// <summary>Non-null (and user-facing) only when the peer reported the listing was
    /// capped — never fabricated when the peer says the listing is complete.</summary>
    public string? TruncationNotice
    {
        get => _truncationNotice;
        private set
        {
            if (SetProperty(ref _truncationNotice, value))
                OnPropertyChanged(nameof(IsTruncated));
        }
    }

    /// <summary>Mirrors <see cref="TruncationNotice"/> as a plain bool, since x:Bind's
    /// InfoBar.IsOpen binding needs a bool rather than a string-emptiness check.</summary>
    public bool IsTruncated => _truncationNotice is not null;

    public MeridianStateViewState State
    {
        get => _state;
        private set => SetProperty(ref _state, value);
    }

    /// <summary>§15-voice: direct, no apology, names the next step. Set only when
    /// <see cref="State"/> is Error.</summary>
    public string? ErrorMessage
    {
        get => _errorMessage;
        private set => SetProperty(ref _errorMessage, value);
    }

    /// <summary>Loads the given remote folder as the new current folder, replacing the
    /// breadcrumb stack entirely (used for the initial load and any direct navigation).</summary>
    public async Task LoadAsync(string path)
    {
        _pathSegments.Clear();
        _pathSegments.AddRange(SplitSegments(path));
        Breadcrumbs = _pathSegments.ToList();

        await FetchAsync(path);
    }

    /// <summary>Navigates into a child folder, pushing one breadcrumb segment.</summary>
    public async Task NavigateIntoAsync(FileEntry entry)
    {
        if (!entry.IsDirectory) return;

        _pathSegments.Add(entry.Name);
        Breadcrumbs = _pathSegments.ToList();

        await FetchAsync(entry.Path);
    }

    /// <summary>Pops one breadcrumb segment and reloads the parent folder. No-op at the root.</summary>
    public async Task NavigateBackAsync()
    {
        if (_pathSegments.Count <= 1) return;

        _pathSegments.RemoveAt(_pathSegments.Count - 1);
        Breadcrumbs = _pathSegments.ToList();

        await FetchAsync(JoinSegments(_pathSegments));
    }

    public void SelectFilter(BrowseFilter filter)
    {
        SelectedFilter = filter;
        Refresh();
    }

    /// <summary>Narrows the current folder's listing by name, client-side — the folder is
    /// already fully (or truncated-but-fully) loaded, so this needs no server round trip.</summary>
    public void SetSearchText(string text)
    {
        _searchText = text ?? string.Empty;
        Refresh();
    }

    private void Refresh()
    {
        var filtered = ApplyFilterAndSort(_allEntries, SelectedFilter);
        Entries = string.IsNullOrWhiteSpace(_searchText)
            ? filtered
            : filtered.Where(e => e.Name.Contains(_searchText, StringComparison.OrdinalIgnoreCase)).ToList();
    }

    private async Task FetchAsync(string path)
    {
        State = MeridianStateViewState.Loading;
        ErrorMessage = null;

        ListResult result;
        try
        {
            result = await _peerHost.ListAsync(path, CancellationToken.None);
        }
        catch (Exception)
        {
            _allEntries = [];
            Entries = [];
            TruncationNotice = null;
            ErrorMessage = "Couldn't load this folder. Try again.";
            State = MeridianStateViewState.Error;
            return;
        }

        _allEntries = result.Entries;
        Refresh();
        TruncationNotice = result.Truncated
            ? $"Showing the first {FileBrowser.MaxEntries:N0} items in this folder."
            : null;

        State = Entries.Count == 0 ? MeridianStateViewState.Empty : MeridianStateViewState.Content;
    }

    private static IReadOnlyList<FileEntry> ApplyFilterAndSort(IReadOnlyList<FileEntry> entries, BrowseFilter filter)
    {
        var filtered = filter == BrowseFilter.All
            ? entries
            : entries.Where(e => e.IsDirectory || Matches(e.Mime, filter)).ToList();

        return filtered
            .OrderBy(e => e.IsDirectory ? 0 : 1)
            .ThenBy(e => e.Name, StringComparer.OrdinalIgnoreCase)
            .ToList();
    }

    private static bool Matches(string? mime, BrowseFilter filter)
    {
        if (mime is null) return false;

        return filter switch
        {
            BrowseFilter.Video => mime.StartsWith("video/", StringComparison.OrdinalIgnoreCase),
            BrowseFilter.Audio => mime.StartsWith("audio/", StringComparison.OrdinalIgnoreCase),
            BrowseFilter.Images => mime.StartsWith("image/", StringComparison.OrdinalIgnoreCase),
            BrowseFilter.Docs => mime.StartsWith("application/", StringComparison.OrdinalIgnoreCase)
                || mime.StartsWith("text/", StringComparison.OrdinalIgnoreCase),
            _ => true,
        };
    }

    private static IEnumerable<string> SplitSegments(string path) =>
        path.Split(['/', '\\'], StringSplitOptions.RemoveEmptyEntries);

    private static string JoinSegments(IReadOnlyList<string> segments) =>
        "/" + string.Join('/', segments);
}

File: android/app/src/main/kotlin/com/slipstream/app/ui/browse/BrowseViewModel.kt
package com.slipstream.app.ui.browse

import androidx.lifecycle.ViewModel
import com.slipstream.app.peer.PeerController
import com.slipstream.meridian.component.MeridianUiState
import java.util.Locale
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/** The mutually-exclusive category filters shown as chips above the listing (Task 6 brief).
 * [matches] is the MIME-prefix predicate each chip filters by — directories are never subject
 * to it (see [BrowseViewModel.applyFilter]), since hiding a folder because its own MIME type
 * isn't a video would make it impossible to navigate into it to find the videos inside. */
enum class BrowseFilter(val label: String) {
    All("All") {
        override fun matches(mime: String?) = true
    },
    Video("Video") {
        override fun matches(mime: String?) = mime?.startsWith("video/") == true
    },
    Audio("Audio") {
        override fun matches(mime: String?) = mime?.startsWith("audio/") == true
    },
    Images("Images") {
        override fun matches(mime: String?) = mime?.startsWith("image/") == true
    },
    Docs("Docs") {
        override fun matches(mime: String?) =
            mime != null && (mime.startsWith("application/") || mime.startsWith("text/"))
    },
    ;

    abstract fun matches(mime: String?): Boolean
}

/** One row in the browse listing — a `:core` [com.slipstream.core.files.FileEntry] plus a
 * ready-to-load `/thumb/<token>` URL (already resolved against the peer's control endpoint by
 * [PeerController.thumbnailUrl], never built here) when the peer supplied one. */
data class BrowseEntry(
    val name: String,
    val isDirectory: Boolean,
    val size: Long,
    val mtimeMs: Long,
    val mime: String?,
    val thumbnailUrl: String?,
)

/** One link in the path trail above the listing. [path] is what gets sent back to
 * [PeerController.list] when the user taps it. */
data class Breadcrumb(val label: String, val path: String)

/** What the browse screen renders. */
data class BrowseState(
    val uiState: MeridianUiState = MeridianUiState.Loading,
    val currentPath: String = "",
    val breadcrumbs: List<Breadcrumb> = emptyList(),
    val filter: BrowseFilter = BrowseFilter.All,
    val entries: List<BrowseEntry> = emptyList(),
    /** All entries for [currentPath], before [filter] narrows them — kept so switching filters
     * never needs a fresh round trip to the peer. */
    val allEntries: List<BrowseEntry> = emptyList(),
    /** Set only when the peer's listing was capped (design.md's directory-entry cap) — an
     * honest "this isn't everything" notice rather than silently pretending it's complete. */
    val truncationNotice: String? = null,
    /** Set once [BrowseViewModel.playHere] has a URL to hand a local player — the screen shows
     * a Media3 player for exactly as long as this is non-null (design.md §8, "Play here"). */
    val playbackUrl: String? = null,
    /** Set when either [BrowseViewModel.playOnPeer] or [BrowseViewModel.playHere] failed — the
     * direct, no-apology message (spec §15), never a stack trace. */
    val playbackError: String? = null,
)

/**
 * Task 6: the browse screen's view model. Lists a directory on the paired peer, sorts
 * directories before files, filters by MIME-prefix chip, and tracks a breadcrumb trail so the
 * user can step back out — all driven purely through [PeerController], `:app`'s one boundary
 * onto `:core`.
 */
class BrowseViewModel(
    private val controller: PeerController,
) : ViewModel() {
    private val _state = MutableStateFlow(BrowseState())
    val state: StateFlow<BrowseState> = _state.asStateFlow()

    /** Lists [path] on the peer and replaces the current screen state with the result — used
     * both for the initial load and for every breadcrumb/directory navigation, since each is
     * just "list a (possibly different) path". */
    suspend fun load(path: String) {
        _state.value = _state.value.copy(uiState = MeridianUiState.Loading)
        val result = controller.list(path)
        result.fold(
            onSuccess = { listing ->
                val entries = listing.entries.map { e ->
                    BrowseEntry(
                        name = e.name,
                        isDirectory = e.isDirectory,
                        size = e.size,
                        mtimeMs = e.mtimeMs,
                        mime = e.mime,
                        thumbnailUrl = e.thumbnailToken?.let { controller.thumbnailUrl(it) },
                    )
                }.sortedWith(compareByDescending<BrowseEntry> { it.isDirectory }.thenBy { it.name.lowercase() })

                val truncationNotice = if (listing.truncated) {
                    String.format(Locale.US, "Showing the first %,d items in this folder.", listing.entries.size)
                } else {
                    null
                }

                _state.value = _state.value.copy(
                    uiState = if (entries.isEmpty()) {
                        MeridianUiState.Empty("This folder is empty.")
                    } else {
                        MeridianUiState.Content
                    },
                    currentPath = path,
                    breadcrumbs = breadcrumbsFor(path),
                    allEntries = entries,
                    entries = applyFilter(entries, _state.value.filter),
                    truncationNotice = truncationNotice,
                )
            },
            onFailure = { error ->
                _state.value = _state.value.copy(
                    uiState = MeridianUiState.Error(
                        message = error.message ?: "Something went wrong.",
                        onRetry = null,
                    ),
                )
            },
        )
    }

    /** Descends into a directory entry, appending it to [BrowseState.currentPath]. */
    suspend fun open(entry: BrowseEntry) {
        require(entry.isDirectory) { "open() is only for directory entries" }
        val base = _state.value.currentPath.trimEnd('/')
        load("$base/${entry.name}")
    }

    /** Jumps back to an earlier breadcrumb, discarding everything below it. */
    suspend fun navigateTo(breadcrumb: Breadcrumb) {
        load(breadcrumb.path)
    }

    /** Narrows (or widens) the visible listing by MIME-prefix category. No round trip to the
     * peer — [BrowseState.allEntries] already holds everything for [BrowseState.currentPath]. */
    fun setFilter(filter: BrowseFilter) {
        _state.value = _state.value.copy(
            filter = filter,
            entries = applyFilter(_state.value.allEntries, filter),
        )
    }

    /**
     * "Play on PC" (design.md §8, push-to-play): [localPath] is a file on *this* device (an
     * absolute path, never resolved against any root — see [PeerController.streamOnPeer]'s
     * doc). Delegates entirely to the controller, which issues this device's own stream token
     * and sends the peer a single `play` message carrying a URL to this device's own media
     * server — nothing is ever copied to the peer.
     */
    suspend fun playOnPeer(localPath: String) {
        val result = controller.streamOnPeer(localPath)
        result.fold(
            onSuccess = {
                _state.value = _state.value.copy(playbackError = null)
            },
            onFailure = { error ->
                _state.value = _state.value.copy(playbackError = error.message ?: "Couldn't start playback.")
            },
        )
    }

    /** "Play here" (design.md §8): [remotePath] is a file on the peer. Fetches a
     * `/media/{token}` URL to hand to a local Media3 player — never downloads the file. */
    suspend fun playHere(remotePath: String) {
        val result = controller.streamUrlFor(remotePath)
        result.fold(
            onSuccess = { url -> _state.value = _state.value.copy(playbackUrl = url, playbackError = null) },
            onFailure = { error ->
                _state.value = _state.value.copy(
                    playbackUrl = null,
                    playbackError = error.message ?: "Couldn't start playback.",
                )
            },
        )
    }

    /** Closes the local player (or clears a playback error banner) shown for [playHere]. */
    fun dismissPlayback() {
        _state.value = _state.value.copy(playbackUrl = null, playbackError = null)
    }

    private fun applyFilter(entries: List<BrowseEntry>, filter: BrowseFilter): List<BrowseEntry> =
        if (filter == BrowseFilter.All) {
            entries
        } else {
            entries.filter { it.isDirectory || filter.matches(it.mime) }
        }

    /** Splits [path] into a breadcrumb per segment, each carrying the full path up to and
     * including itself, e.g. `/root/Photos` -> `[("root", "/root"), ("Photos", "/root/Photos")]`. */
    private fun breadcrumbsFor(path: String): List<Breadcrumb> {
        val segments = path.trim('/').split('/').filter { it.isNotBlank() }
        if (segments.isEmpty()) return listOf(Breadcrumb(label = "Root", path = "/"))
        var accumulated = ""
        return segments.map { segment ->
            accumulated += "/$segment"
            Breadcrumb(label = segment, path = accumulated)
        }
    }
}
